<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>An Introduction to Machine Learning</title>
  <meta name="description" content="Course materials for An Introduction to Machine Learning">
  <meta name="generator" content="bookdown 0.7 and GitBook 2.6.7">

  <meta property="og:title" content="An Introduction to Machine Learning" />
  <meta property="og:type" content="book" />
  
  <meta property="og:image" content="figures/cover_image.png" />
  <meta property="og:description" content="Course materials for An Introduction to Machine Learning" />
  <meta name="github-repo" content="bioinformatics-training/intro-machine-learning-2018" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="An Introduction to Machine Learning" />
  
  <meta name="twitter:description" content="Course materials for An Introduction to Machine Learning" />
  <meta name="twitter:image" content="figures/cover_image.png" />

<meta name="author" content="Sudhakaran Prabakaran, Matt Wayland and Chris Penfold">


<meta name="date" content="2018-09-25">

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<link rel="prev" href="logistic-regression.html">
<link rel="next" href="clustering.html">
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />









<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">An Introduction to Machine Learning</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> About the course</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#overview"><i class="fa fa-check"></i><b>1.1</b> Overview</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#registration"><i class="fa fa-check"></i><b>1.2</b> Registration</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#prerequisites"><i class="fa fa-check"></i><b>1.3</b> Prerequisites</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#github"><i class="fa fa-check"></i><b>1.4</b> Github</a></li>
<li class="chapter" data-level="1.5" data-path="index.html"><a href="index.html#license"><i class="fa fa-check"></i><b>1.5</b> License</a></li>
<li class="chapter" data-level="1.6" data-path="index.html"><a href="index.html#contact"><i class="fa fa-check"></i><b>1.6</b> Contact</a></li>
<li class="chapter" data-level="1.7" data-path="index.html"><a href="index.html#colophon"><i class="fa fa-check"></i><b>1.7</b> Colophon</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="intro.html"><a href="intro.html"><i class="fa fa-check"></i><b>2</b> Introduction</a><ul>
<li class="chapter" data-level="2.1" data-path="intro.html"><a href="intro.html#what-is-machine-learning"><i class="fa fa-check"></i><b>2.1</b> What is machine learning?</a></li>
<li class="chapter" data-level="2.2" data-path="intro.html"><a href="intro.html#aspects-of-ml"><i class="fa fa-check"></i><b>2.2</b> Aspects of ML</a></li>
<li class="chapter" data-level="2.3" data-path="intro.html"><a href="intro.html#what-actually-happened-under-the-hood"><i class="fa fa-check"></i><b>2.3</b> What actually happened under the hood</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="linear-models.html"><a href="linear-models.html"><i class="fa fa-check"></i><b>3</b> Linear models and matrix algebra</a><ul>
<li class="chapter" data-level="3.1" data-path="linear-models.html"><a href="linear-models.html#linear-models"><i class="fa fa-check"></i><b>3.1</b> Linear models</a></li>
<li class="chapter" data-level="3.2" data-path="linear-models.html"><a href="linear-models.html#matrix-algebra"><i class="fa fa-check"></i><b>3.2</b> Matrix algebra</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="logistic-regression.html"><a href="logistic-regression.html"><i class="fa fa-check"></i><b>4</b> Supervised learning</a><ul>
<li class="chapter" data-level="4.1" data-path="logistic-regression.html"><a href="logistic-regression.html#regression"><i class="fa fa-check"></i><b>4.1</b> Regression</a><ul>
<li class="chapter" data-level="4.1.1" data-path="logistic-regression.html"><a href="logistic-regression.html#linear-regression"><i class="fa fa-check"></i><b>4.1.1</b> Linear regression</a></li>
<li class="chapter" data-level="4.1.2" data-path="logistic-regression.html"><a href="logistic-regression.html#polynomial-regression"><i class="fa fa-check"></i><b>4.1.2</b> Polynomial regression</a></li>
<li class="chapter" data-level="4.1.3" data-path="logistic-regression.html"><a href="logistic-regression.html#distributions-of-fits"><i class="fa fa-check"></i><b>4.1.3</b> Distributions of fits</a></li>
<li class="chapter" data-level="4.1.4" data-path="logistic-regression.html"><a href="logistic-regression.html#gaussian-process-regression"><i class="fa fa-check"></i><b>4.1.4</b> Gaussian process regression</a></li>
</ul></li>
<li class="chapter" data-level="4.2" data-path="logistic-regression.html"><a href="logistic-regression.html#classification"><i class="fa fa-check"></i><b>4.2</b> Classification</a><ul>
<li class="chapter" data-level="4.2.1" data-path="logistic-regression.html"><a href="logistic-regression.html#logistic-regression"><i class="fa fa-check"></i><b>4.2.1</b> Logistic regression</a></li>
<li class="chapter" data-level="4.2.2" data-path="logistic-regression.html"><a href="logistic-regression.html#gp-classification"><i class="fa fa-check"></i><b>4.2.2</b> GP classification</a></li>
<li class="chapter" data-level="4.2.3" data-path="logistic-regression.html"><a href="logistic-regression.html#other-classification-approaches."><i class="fa fa-check"></i><b>4.2.3</b> Other classification approaches.</a></li>
</ul></li>
<li class="chapter" data-level="4.3" data-path="logistic-regression.html"><a href="logistic-regression.html#resources"><i class="fa fa-check"></i><b>4.3</b> Resources</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html"><i class="fa fa-check"></i><b>5</b> Dimensionality reduction</a><ul>
<li class="chapter" data-level="5.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#linear-dimensionality-reduction"><i class="fa fa-check"></i><b>5.1</b> Linear Dimensionality Reduction</a><ul>
<li class="chapter" data-level="5.1.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#interpreting-the-principle-component-axes"><i class="fa fa-check"></i><b>5.1.1</b> Interpreting the Principle Component Axes</a></li>
<li class="chapter" data-level="5.1.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#horseshoe-effect"><i class="fa fa-check"></i><b>5.1.2</b> Horseshoe effect</a></li>
<li class="chapter" data-level="5.1.3" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#pca-analysis-of-mammalian-development"><i class="fa fa-check"></i><b>5.1.3</b> PCA analysis of mammalian development</a></li>
</ul></li>
<li class="chapter" data-level="5.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#nonlinear-dimensionality-reduction"><i class="fa fa-check"></i><b>5.2</b> Nonlinear Dimensionality Reduction</a><ul>
<li class="chapter" data-level="5.2.1" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#nonlinear-warping"><i class="fa fa-check"></i><b>5.2.1</b> Nonlinear warping</a></li>
<li class="chapter" data-level="5.2.2" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#stochasticity"><i class="fa fa-check"></i><b>5.2.2</b> Stochasticity</a></li>
<li class="chapter" data-level="5.2.3" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#analysis-of-mammalian-development"><i class="fa fa-check"></i><b>5.2.3</b> Analysis of mammalian development</a></li>
</ul></li>
<li class="chapter" data-level="5.3" data-path="dimensionality-reduction.html"><a href="dimensionality-reduction.html#other-dimensionality-reduction-techniques"><i class="fa fa-check"></i><b>5.3</b> Other dimensionality reduction techniques</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>6</b> Clustering</a><ul>
<li class="chapter" data-level="6.1" data-path="clustering.html"><a href="clustering.html#introduction"><i class="fa fa-check"></i><b>6.1</b> Introduction</a></li>
<li class="chapter" data-level="6.2" data-path="clustering.html"><a href="clustering.html#distance-metrics"><i class="fa fa-check"></i><b>6.2</b> Distance metrics</a></li>
<li class="chapter" data-level="6.3" data-path="clustering.html"><a href="clustering.html#hierarchic-agglomerative"><i class="fa fa-check"></i><b>6.3</b> Hierarchic agglomerative</a><ul>
<li class="chapter" data-level="6.3.1" data-path="clustering.html"><a href="clustering.html#linkage-algorithms"><i class="fa fa-check"></i><b>6.3.1</b> Linkage algorithms</a></li>
<li class="chapter" data-level="6.3.2" data-path="clustering.html"><a href="clustering.html#example-clustering-synthetic-data-sets"><i class="fa fa-check"></i><b>6.3.2</b> Example: clustering synthetic data sets</a></li>
<li class="chapter" data-level="6.3.3" data-path="clustering.html"><a href="clustering.html#example-gene-expression-profiling-of-human-tissues"><i class="fa fa-check"></i><b>6.3.3</b> Example: gene expression profiling of human tissues</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="clustering.html"><a href="clustering.html#k-means"><i class="fa fa-check"></i><b>6.4</b> K-means</a><ul>
<li class="chapter" data-level="6.4.1" data-path="clustering.html"><a href="clustering.html#algorithm"><i class="fa fa-check"></i><b>6.4.1</b> Algorithm</a></li>
<li class="chapter" data-level="6.4.2" data-path="clustering.html"><a href="clustering.html#choosing-initial-cluster-centres"><i class="fa fa-check"></i><b>6.4.2</b> Choosing initial cluster centres</a></li>
<li class="chapter" data-level="6.4.3" data-path="clustering.html"><a href="clustering.html#choosingK"><i class="fa fa-check"></i><b>6.4.3</b> Choosing k</a></li>
<li class="chapter" data-level="6.4.4" data-path="clustering.html"><a href="clustering.html#example-clustering-synthetic-data-sets-1"><i class="fa fa-check"></i><b>6.4.4</b> Example: clustering synthetic data sets</a></li>
<li class="chapter" data-level="6.4.5" data-path="clustering.html"><a href="clustering.html#example-gene-expression-profiling-of-human-tissues-1"><i class="fa fa-check"></i><b>6.4.5</b> Example: gene expression profiling of human tissues</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="clustering.html"><a href="clustering.html#dbscan"><i class="fa fa-check"></i><b>6.5</b> DBSCAN</a><ul>
<li class="chapter" data-level="6.5.1" data-path="clustering.html"><a href="clustering.html#algorithm-1"><i class="fa fa-check"></i><b>6.5.1</b> Algorithm</a></li>
<li class="chapter" data-level="6.5.2" data-path="clustering.html"><a href="clustering.html#implementation-in-r"><i class="fa fa-check"></i><b>6.5.2</b> Implementation in R</a></li>
<li class="chapter" data-level="6.5.3" data-path="clustering.html"><a href="clustering.html#choosing-parameters"><i class="fa fa-check"></i><b>6.5.3</b> Choosing parameters</a></li>
<li class="chapter" data-level="6.5.4" data-path="clustering.html"><a href="clustering.html#example-clustering-synthetic-data-sets-2"><i class="fa fa-check"></i><b>6.5.4</b> Example: clustering synthetic data sets</a></li>
<li class="chapter" data-level="6.5.5" data-path="clustering.html"><a href="clustering.html#example-gene-expression-profiling-of-human-tissues-2"><i class="fa fa-check"></i><b>6.5.5</b> Example: gene expression profiling of human tissues</a></li>
</ul></li>
<li class="chapter" data-level="6.6" data-path="clustering.html"><a href="clustering.html#evaluating-cluster-quality"><i class="fa fa-check"></i><b>6.6</b> Evaluating cluster quality</a><ul>
<li class="chapter" data-level="6.6.1" data-path="clustering.html"><a href="clustering.html#silhouetteMethod"><i class="fa fa-check"></i><b>6.6.1</b> Silhouette method</a></li>
<li class="chapter" data-level="6.6.2" data-path="clustering.html"><a href="clustering.html#example---k-means-clustering-of-blobs-data-set"><i class="fa fa-check"></i><b>6.6.2</b> Example - k-means clustering of blobs data set</a></li>
<li class="chapter" data-level="6.6.3" data-path="clustering.html"><a href="clustering.html#example---dbscan-clustering-of-noisy-moons"><i class="fa fa-check"></i><b>6.6.3</b> Example - DBSCAN clustering of noisy moons</a></li>
</ul></li>
<li class="chapter" data-level="6.7" data-path="clustering.html"><a href="clustering.html#exercises"><i class="fa fa-check"></i><b>6.7</b> Exercises</a><ul>
<li class="chapter" data-level="6.7.1" data-path="clustering.html"><a href="clustering.html#clusteringEx1"><i class="fa fa-check"></i><b>6.7.1</b> Exercise 1</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html"><i class="fa fa-check"></i><b>7</b> Nearest neighbours</a><ul>
<li class="chapter" data-level="7.1" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#introduction-1"><i class="fa fa-check"></i><b>7.1</b> Introduction</a></li>
<li class="chapter" data-level="7.2" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#classification-simulated-data"><i class="fa fa-check"></i><b>7.2</b> Classification: simulated data</a><ul>
<li class="chapter" data-level="7.2.1" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#knn-function"><i class="fa fa-check"></i><b>7.2.1</b> knn function</a></li>
<li class="chapter" data-level="7.2.2" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#plotting-decision-boundaries"><i class="fa fa-check"></i><b>7.2.2</b> Plotting decision boundaries</a></li>
<li class="chapter" data-level="7.2.3" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#bias-variance-tradeoff"><i class="fa fa-check"></i><b>7.2.3</b> Bias-variance tradeoff</a></li>
<li class="chapter" data-level="7.2.4" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#choosing-k"><i class="fa fa-check"></i><b>7.2.4</b> Choosing <em>k</em></a></li>
</ul></li>
<li class="chapter" data-level="7.3" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#knn-cell-segmentation"><i class="fa fa-check"></i><b>7.3</b> Classification: cell segmentation</a><ul>
<li class="chapter" data-level="7.3.1" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#cell-segmentation-data-set"><i class="fa fa-check"></i><b>7.3.1</b> Cell segmentation data set</a></li>
<li class="chapter" data-level="7.3.2" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#data-splitting"><i class="fa fa-check"></i><b>7.3.2</b> Data splitting</a></li>
<li class="chapter" data-level="7.3.3" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#identification-of-data-quality-issues"><i class="fa fa-check"></i><b>7.3.3</b> Identification of data quality issues</a></li>
<li class="chapter" data-level="7.3.4" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#fit-model"><i class="fa fa-check"></i><b>7.3.4</b> Fit model</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#knn-regression"><i class="fa fa-check"></i><b>7.4</b> Regression</a><ul>
<li class="chapter" data-level="7.4.1" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#partition-data"><i class="fa fa-check"></i><b>7.4.1</b> Partition data</a></li>
<li class="chapter" data-level="7.4.2" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#data-pre-processing"><i class="fa fa-check"></i><b>7.4.2</b> Data pre-processing</a></li>
<li class="chapter" data-level="7.4.3" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#search-for-optimum-k"><i class="fa fa-check"></i><b>7.4.3</b> Search for optimum <em>k</em></a></li>
<li class="chapter" data-level="7.4.4" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#use-model-to-make-predictions"><i class="fa fa-check"></i><b>7.4.4</b> Use model to make predictions</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#exercises-1"><i class="fa fa-check"></i><b>7.5</b> Exercises</a><ul>
<li class="chapter" data-level="7.5.1" data-path="nearest-neighbours.html"><a href="nearest-neighbours.html#knnEx1"><i class="fa fa-check"></i><b>7.5.1</b> Exercise 1</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="svm.html"><a href="svm.html"><i class="fa fa-check"></i><b>8</b> Support vector machines</a><ul>
<li class="chapter" data-level="8.1" data-path="svm.html"><a href="svm.html#introduction-2"><i class="fa fa-check"></i><b>8.1</b> Introduction</a><ul>
<li class="chapter" data-level="8.1.1" data-path="svm.html"><a href="svm.html#maximum-margin-classifier"><i class="fa fa-check"></i><b>8.1.1</b> Maximum margin classifier</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="svm.html"><a href="svm.html#support-vector-classifier"><i class="fa fa-check"></i><b>8.2</b> Support vector classifier</a></li>
<li class="chapter" data-level="8.3" data-path="svm.html"><a href="svm.html#support-vector-machine"><i class="fa fa-check"></i><b>8.3</b> Support Vector Machine</a></li>
<li class="chapter" data-level="8.4" data-path="svm.html"><a href="svm.html#example---training-a-classifier"><i class="fa fa-check"></i><b>8.4</b> Example - training a classifier</a><ul>
<li class="chapter" data-level="8.4.1" data-path="svm.html"><a href="svm.html#setup-environment"><i class="fa fa-check"></i><b>8.4.1</b> Setup environment</a></li>
<li class="chapter" data-level="8.4.2" data-path="svm.html"><a href="svm.html#partition-data-1"><i class="fa fa-check"></i><b>8.4.2</b> Partition data</a></li>
<li class="chapter" data-level="8.4.3" data-path="svm.html"><a href="svm.html#visualize-training-data"><i class="fa fa-check"></i><b>8.4.3</b> Visualize training data</a></li>
<li class="chapter" data-level="8.4.4" data-path="svm.html"><a href="svm.html#define-a-custom-model"><i class="fa fa-check"></i><b>8.4.4</b> Define a custom model</a></li>
<li class="chapter" data-level="8.4.5" data-path="svm.html"><a href="svm.html#model-cross-validation-and-tuning"><i class="fa fa-check"></i><b>8.4.5</b> Model cross-validation and tuning</a></li>
<li class="chapter" data-level="8.4.6" data-path="svm.html"><a href="svm.html#prediction-performance-measures"><i class="fa fa-check"></i><b>8.4.6</b> Prediction performance measures</a></li>
<li class="chapter" data-level="8.4.7" data-path="svm.html"><a href="svm.html#plot-decision-boundary"><i class="fa fa-check"></i><b>8.4.7</b> Plot decision boundary</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="svm.html"><a href="svm.html#example---regression"><i class="fa fa-check"></i><b>8.5</b> Example - regression</a></li>
<li class="chapter" data-level="8.6" data-path="svm.html"><a href="svm.html#further-reading"><i class="fa fa-check"></i><b>8.6</b> Further reading</a></li>
<li class="chapter" data-level="8.7" data-path="svm.html"><a href="svm.html#exercises-2"><i class="fa fa-check"></i><b>8.7</b> Exercises</a><ul>
<li class="chapter" data-level="8.7.1" data-path="svm.html"><a href="svm.html#exercise-1"><i class="fa fa-check"></i><b>8.7.1</b> Exercise 1</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="9" data-path="decision-trees.html"><a href="decision-trees.html"><i class="fa fa-check"></i><b>9</b> Decision trees and random forests</a><ul>
<li class="chapter" data-level="9.1" data-path="decision-trees.html"><a href="decision-trees.html#decision-trees"><i class="fa fa-check"></i><b>9.1</b> Decision Trees</a></li>
<li class="chapter" data-level="9.2" data-path="decision-trees.html"><a href="decision-trees.html#random-forest"><i class="fa fa-check"></i><b>9.2</b> Random Forest</a></li>
<li class="chapter" data-level="9.3" data-path="decision-trees.html"><a href="decision-trees.html#exercises-3"><i class="fa fa-check"></i><b>9.3</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="ann.html"><a href="ann.html"><i class="fa fa-check"></i><b>10</b> Artificial neural networks</a><ul>
<li class="chapter" data-level="10.1" data-path="ann.html"><a href="ann.html#neural-networks"><i class="fa fa-check"></i><b>10.1</b> Neural Networks</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="mlnn.html"><a href="mlnn.html"><i class="fa fa-check"></i><b>11</b> Deep Learning</a><ul>
<li class="chapter" data-level="11.1" data-path="mlnn.html"><a href="mlnn.html#multilayer-neural-networks"><i class="fa fa-check"></i><b>11.1</b> Multilayer Neural Networks</a><ul>
<li class="chapter" data-level="11.1.1" data-path="mlnn.html"><a href="mlnn.html#constructing-layers-in-kerasr"><i class="fa fa-check"></i><b>11.1.1</b> Constructing layers in kerasR</a></li>
<li class="chapter" data-level="11.1.2" data-path="mlnn.html"><a href="mlnn.html#reading-in-images"><i class="fa fa-check"></i><b>11.1.2</b> Reading in images</a></li>
<li class="chapter" data-level="11.1.3" data-path="mlnn.html"><a href="mlnn.html#rick-and-morty-classifier-using-deep-learning"><i class="fa fa-check"></i><b>11.1.3</b> Rick and Morty classifier using Deep Learning</a></li>
</ul></li>
<li class="chapter" data-level="11.2" data-path="mlnn.html"><a href="mlnn.html#convolutional-neural-networks"><i class="fa fa-check"></i><b>11.2</b> Convolutional neural networks</a><ul>
<li class="chapter" data-level="11.2.1" data-path="mlnn.html"><a href="mlnn.html#data-augmentation"><i class="fa fa-check"></i><b>11.2.1</b> Data augmentation</a></li>
<li class="chapter" data-level="11.2.2" data-path="mlnn.html"><a href="mlnn.html#asking-more-precise-questions"><i class="fa fa-check"></i><b>11.2.2</b> Asking more precise questions</a></li>
<li class="chapter" data-level="11.2.3" data-path="mlnn.html"><a href="mlnn.html#more-complex-networks"><i class="fa fa-check"></i><b>11.2.3</b> More complex networks</a></li>
</ul></li>
<li class="chapter" data-level="11.3" data-path="mlnn.html"><a href="mlnn.html#further-reading-1"><i class="fa fa-check"></i><b>11.3</b> Further reading</a></li>
</ul></li>
<li class="appendix"><span><b>Appendix</b></span></li>
<li class="chapter" data-level="A" data-path="logistic-regression.html"><a href="logistic-regression.html#resources"><i class="fa fa-check"></i><b>A</b> Resources</a><ul>
<li class="chapter" data-level="A.1" data-path="resources.html"><a href="resources.html"><i class="fa fa-check"></i><b>A.1</b> Python</a></li>
<li class="chapter" data-level="A.2" data-path="resources.html"><a href="resources.html#machine-learning-data-set-repositories"><i class="fa fa-check"></i><b>A.2</b> Machine learning data set repositories</a><ul>
<li class="chapter" data-level="A.2.1" data-path="resources.html"><a href="resources.html#mldata"><i class="fa fa-check"></i><b>A.2.1</b> MLDATA</a></li>
<li class="chapter" data-level="A.2.2" data-path="resources.html"><a href="resources.html#uci-machine-learning-repository"><i class="fa fa-check"></i><b>A.2.2</b> UCI Machine Learning Repository</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="B" data-path="solutions-linear-models.html"><a href="solutions-linear-models.html"><i class="fa fa-check"></i><b>B</b> Solutions ch. 3 - Linear models and matrix algebra</a><ul>
<li class="chapter" data-level="B.1" data-path="solutions-linear-models.html"><a href="solutions-linear-models.html#example-2"><i class="fa fa-check"></i><b>B.1</b> Example 2</a></li>
<li class="chapter" data-level="B.2" data-path="solutions-linear-models.html"><a href="solutions-linear-models.html#example-2-1"><i class="fa fa-check"></i><b>B.2</b> Example 2</a></li>
</ul></li>
<li class="chapter" data-level="C" data-path="solutions-logistic-regression.html"><a href="solutions-logistic-regression.html"><i class="fa fa-check"></i><b>C</b> Solutions ch. 4 - Linear and non-linear (logistic) regression</a></li>
<li class="chapter" data-level="D" data-path="solutions-dimensionality-reduction.html"><a href="solutions-dimensionality-reduction.html"><i class="fa fa-check"></i><b>D</b> Solutions ch. 5 - Dimensionality reduction</a><ul>
<li class="chapter" data-level="D.1" data-path="solutions-dimensionality-reduction.html"><a href="solutions-dimensionality-reduction.html#exercise-5.1"><i class="fa fa-check"></i><b>D.1</b> Exercise 5.1</a></li>
<li class="chapter" data-level="D.2" data-path="solutions-dimensionality-reduction.html"><a href="solutions-dimensionality-reduction.html#exercise-5.2"><i class="fa fa-check"></i><b>D.2</b> Exercise 5.2</a></li>
<li class="chapter" data-level="D.3" data-path="solutions-dimensionality-reduction.html"><a href="solutions-dimensionality-reduction.html#exercise-5.3."><i class="fa fa-check"></i><b>D.3</b> Exercise 5.3.</a></li>
<li class="chapter" data-level="D.4" data-path="solutions-dimensionality-reduction.html"><a href="solutions-dimensionality-reduction.html#exercise-5.4."><i class="fa fa-check"></i><b>D.4</b> Exercise 5.4.</a></li>
<li class="chapter" data-level="D.5" data-path="solutions-dimensionality-reduction.html"><a href="solutions-dimensionality-reduction.html#exercise-5.5"><i class="fa fa-check"></i><b>D.5</b> Exercise 5.5</a></li>
<li class="chapter" data-level="D.6" data-path="solutions-dimensionality-reduction.html"><a href="solutions-dimensionality-reduction.html#exercise-5.6."><i class="fa fa-check"></i><b>D.6</b> Exercise 5.6.</a></li>
<li class="chapter" data-level="D.7" data-path="solutions-dimensionality-reduction.html"><a href="solutions-dimensionality-reduction.html#exercise-5.7."><i class="fa fa-check"></i><b>D.7</b> Exercise 5.7.</a></li>
<li class="chapter" data-level="D.8" data-path="solutions-dimensionality-reduction.html"><a href="solutions-dimensionality-reduction.html#exercise-5.8."><i class="fa fa-check"></i><b>D.8</b> Exercise 5.8.</a></li>
</ul></li>
<li class="chapter" data-level="E" data-path="solutions-clustering.html"><a href="solutions-clustering.html"><i class="fa fa-check"></i><b>E</b> Solutions ch. 6 - Clustering</a><ul>
<li class="chapter" data-level="E.1" data-path="solutions-clustering.html"><a href="solutions-clustering.html#exercise-1-1"><i class="fa fa-check"></i><b>E.1</b> Exercise 1</a></li>
</ul></li>
<li class="chapter" data-level="F" data-path="solutions-nearest-neighbours.html"><a href="solutions-nearest-neighbours.html"><i class="fa fa-check"></i><b>F</b> Solutions ch. 7 - Nearest neighbours</a><ul>
<li class="chapter" data-level="F.1" data-path="solutions-nearest-neighbours.html"><a href="solutions-nearest-neighbours.html#exercise-1-2"><i class="fa fa-check"></i><b>F.1</b> Exercise 1</a></li>
</ul></li>
<li class="chapter" data-level="G" data-path="solutions-svm.html"><a href="solutions-svm.html"><i class="fa fa-check"></i><b>G</b> Solutions ch. 8 - Support vector machines</a><ul>
<li class="chapter" data-level="G.1" data-path="solutions-svm.html"><a href="solutions-svm.html#exercise-1-3"><i class="fa fa-check"></i><b>G.1</b> Exercise 1</a></li>
</ul></li>
<li class="chapter" data-level="H" data-path="solutions-decision-trees.html"><a href="solutions-decision-trees.html"><i class="fa fa-check"></i><b>H</b> Solutions ch. 9 - Decision trees and random forests</a><ul>
<li class="chapter" data-level="H.1" data-path="solutions-decision-trees.html"><a href="solutions-decision-trees.html#exercise-1-4"><i class="fa fa-check"></i><b>H.1</b> Exercise 1</a></li>
</ul></li>
<li class="chapter" data-level="I" data-path="solutions-ann.html"><a href="solutions-ann.html"><i class="fa fa-check"></i><b>I</b> Solutions ch. 10 - Artificial neural networks</a><ul>
<li class="chapter" data-level="I.1" data-path="solutions-ann.html"><a href="solutions-ann.html#exercise-1-5"><i class="fa fa-check"></i><b>I.1</b> Exercise 1</a></li>
</ul></li>
<li class="chapter" data-level="" data-path="references.html"><a href="references.html"><i class="fa fa-check"></i>References</a></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">An Introduction to Machine Learning</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="dimensionality-reduction" class="section level1">
<h1><span class="header-section-number">5</span> Dimensionality reduction</h1>
<p>In machine learning, dimensionality reduction broadly refers to any modelling approach that reduces the number of variables in a dataset to a few highly informative or representative ones (see Figure <a href="dimensionality-reduction.html#fig:dimreduc">5.1</a>). This is necessitated by the fact that large datasets with many variables are inherently difficult for humans to develop a clear intuition for. Dimensionality reduction is therefore an integral step in the analysis of large, complex biological datasets, allowing exploratory analyses and more intuitive visualisation that may aid interpretability.</p>
<div class="figure" style="text-align: center"><span id="fig:dimreduc"></span>
<img src="images/swiss_roll_manifold_sculpting.png" alt="Example of a dimensionality reduction. Here we have a two-dimensional dataset embeded in a three-dimensional space (swiss roll dataset)." width="55%" />
<p class="caption">
Figure 5.1: Example of a dimensionality reduction. Here we have a two-dimensional dataset embeded in a three-dimensional space (swiss roll dataset).
</p>
</div>
<p>In biological applications, systems-level measurements are typically used to decipher complex mechanisms. These include measurements of gene expression from collections of microarrays <span class="citation">(Breeze et al. <a href="#ref-Breeze873">2011</a>,<span class="citation">Windram et al. (<a href="#ref-windram2012arabidopsis">2012</a>)</span>,<span class="citation">Lewis et al. (<a href="#ref-Lewis15">2015</a>)</span>,<span class="citation">Bechtold et al. (<a href="#ref-Bechtold">2016</a>)</span>)</span> or RNA-sequencing experiments <span class="citation">(Irie et al. <a href="#ref-irie2015sox17">2015</a>,<span class="citation">Tang et al. (<a href="#ref-tang2015unique">2015</a>)</span>)</span> that provide quantitative measurments for tens-of-thousands of genes. Studies like these, based on bulk measurements (that is pooled material), provide observations for many variables (in this case many genes) but with relatively few samples e.g., few time points or conditions. The imbalance between the number of variables and the number of observations is referred to as large <em>p</em>, small <em>n</em>, and makes statistical analysis difficult. Dimensionality reduction techniques therefore prove to be a useful first step in any analysis, identifying potential structure that exists in the dataset or highlighting which (combinations of) variables are the most informative.</p>
<p>The increasing prevalence of single cell RNA-sequencing (scRNA-seq) means the scale of datasets has shifted away from large <em>p</em>, small <em>n</em>, towards providing measurements of many variables but with a corresponding large number of observations (large <em>n</em>) albeit from potentially heterogeneous populations. scRNA-sequencing was largely driven by the need to investigate the transcrptomes of cells that were limited in quantity, such as embryonic cells, with early applications in mouse blastomeres <span class="citation">(F. Tang et al. <a href="#ref-tang2009mrna">2009</a>)</span>. As of 2017, scRNA-seq experiments routinely generate datasets with tens to hundreds-of-thousands of cells (see e.g., <span class="citation">(Svensson, Vento-Tormo, and Teichmann <a href="#ref-svensson2017moore">2017</a>)</span>). Indeed, in 2016, the <a href="https://community.10xgenomics.com/t5/10x-Blog/Our-1-3-million-single-cell-dataset-is-ready-to-download/ba-p/276">10x Genomics million cell experiment</a> provided sequencing for over 1.3 million cells taken from the cortex, hippocampus and ventricular zone of embryonic mice, and large international consortiums, such as the <a href="https://www.humancellatlas.org">Human Cell Atlas</a> aim to create a comprehensive maps of all cell types in the human body. A key goal when dealing with datasets of this magnitude is the identification of subpopulations of cells that may have gone undetected in bulk experiments; another, perhaps more ambitious task, aims to take advantage of any heterogeneity within the population in order to identify a temporal or mechanistic progression of developmental processes or disease.</p>
<p>Of course, whilst dimensionality reduction allows humans to inspect the dataset manually, particularly when the data can be represented in two or three dimensions, we should keep in mind that humans are exceptionally good at identifying patterns in two or three dimensional data, even when no real structure exists (Figure <a href="dimensionality-reduction.html#fig:humanpattern">5.2</a>. It is therefore useful to employ other statistical approaches to search for patterns in the reduced dimensional space. In this sense, dimensionality reduction forms an integral component in the analysis of complex datasets that will typically be combined a variety of machine learning techniques, such as classification, regression, and clustering.</p>
<div class="figure" style="text-align: center"><span id="fig:humanpattern"></span>
<img src="images/GB1.jpg" alt="Humans are exceptionally good at identifying patterns in two and three-dimensional spaces - sometimes too good. To illustrate this, note the Great Britain shapped cloud in the image (presumably drifting away from an EU shaped cloud, not shown). More whimsical shaped clouds can also be seen if you have a spare afternoon.  Golcar Matt/Weatherwatchers [BBC News](http://www.bbc.co.uk/news/uk-england-leeds-40287817)" width="35%" />
<p class="caption">
Figure 5.2: Humans are exceptionally good at identifying patterns in two and three-dimensional spaces - sometimes too good. To illustrate this, note the Great Britain shapped cloud in the image (presumably drifting away from an EU shaped cloud, not shown). More whimsical shaped clouds can also be seen if you have a spare afternoon. Golcar Matt/Weatherwatchers <a href="http://www.bbc.co.uk/news/uk-england-leeds-40287817">BBC News</a>
</p>
</div>
<p>In this chapter we will explore two forms of dimensionality reduction, principle component analysis (<a href="dimensionality-reduction.html#linear-dimensionality-reduction">PCA</a>) and t-distributed stochastic neighbour embedding (<a href="dimensionality-reduction.html#nonlinear-dimensionality-reduction">tSNE</a>), highlighting the advantages and potential pitfalls of each method. As an illustrative example, we will use these approaches to analyse single cell RNA-sequencing data of early human development.</p>
<div id="linear-dimensionality-reduction" class="section level2">
<h2><span class="header-section-number">5.1</span> Linear Dimensionality Reduction</h2>
<p>The most widely used form of dimensionality reduction is principle component analysis (PCA), which was introduced by Pearson in the early 1900’s <span class="citation">(Pearson <a href="#ref-pearson1901liii">1901</a>)</span>, and independently rediscovered by Hotelling <span class="citation">(Hotelling <a href="#ref-hotelling1933analysis">1933</a>)</span>. PCA has a long history of use in biological and ecological applications, with early use in population studies <span class="citation">(Sforza and Edwards <a href="#ref-sforza1964analysis">1964</a>)</span>, and later for the analysis of gene expression data <span class="citation">(Vohradsky, Li, and Thompson <a href="#ref-vohradsky1997identification">1997</a>,<span class="citation">Craig et al. (<a href="#ref-craig1997developmental">1997</a>)</span>,<span class="citation">Hilsenbeck et al. (<a href="#ref-hilsenbeck1999statistical">1999</a>)</span>)</span>.</p>
<p>PCA is not a dimensionality reduction technique <em>per se</em>, but an alternative way of representing the data that more naturally captures the variance in the system. Specifically, it finds a new co-ordinate system, so that the new “x-axis” (which is called the first principle component; PC1) is aligned along the direction of greatest variance, with an orthogonal “y-axis” aligned along the direction with second greatest variance (the second principle component; PC2), and so forth. At this stage there has been no inherent reduction in the dimensionality of the system, we have simply rotated the data around.</p>
<p>To illustrate PCA we can repeat the analysis of <span class="citation">(Ringnér <a href="#ref-ringner2008principal">2008</a>)</span> using the dataset of <span class="citation">(Saal et al. <a href="#ref-saal2007poor">2007</a>)</span> (GEO GSE5325). This dataset contains gene expression profiles for <span class="math inline">\(105\)</span> breast tumour samples measured using Swegene Human 27K RAP UniGene188 arrays. Within the population of cells, <span class="citation">(Ringnér <a href="#ref-ringner2008principal">2008</a>)</span> focused on the expression of <em>GATA3</em> and <em>XBP1</em>, whose expression was known to correlate with estrogen receptor status [^](Breast cancer cells may be estrogen receptor positive, ER<span class="math inline">\(^+\)</span>, or negative, ER<span class="math inline">\(^-\)</span>, indicating capacity to respond to estrogen signalling, which has impliations for treatment), representing a two dimensional system. A pre-processed dataset containing the expression levels for <em>GATA3</em> and <em>XBP1</em>, and ER status, can be loaded into R using the code, below:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">D &lt;-<span class="st"> </span><span class="kw">read.csv</span>(<span class="dt">file =</span> <span class="st">&quot;data/GSE5325/GSE5325_markers.csv&quot;</span>, <span class="dt">header =</span> <span class="ot">TRUE</span>, <span class="dt">sep =</span> <span class="st">&quot;,&quot;</span>, <span class="dt">row.names=</span><span class="dv">1</span>)</code></pre></div>
<p>We can now plot the expression levels of <em>GATA3</em> and <em>XBP1</em> (rows 1 and 2) against one another to visualise the data in the two-dimensional space:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(<span class="kw">t</span>(D[<span class="dv">1</span>,<span class="kw">which</span>(D[<span class="dv">3</span>,]<span class="op">==</span><span class="dv">0</span>)]),<span class="kw">t</span>(D[<span class="dv">2</span>,<span class="kw">which</span>(D[<span class="dv">3</span>,]<span class="op">==</span><span class="dv">0</span>)]),<span class="st">&#39;p&#39;</span>,<span class="dt">col=</span><span class="st">&#39;red&#39;</span>, <span class="dt">ylab=</span><span class="st">&quot;XBP1&quot;</span>, <span class="dt">xlab=</span><span class="st">&quot;GATA3&quot;</span>,<span class="dt">xlim=</span><span class="kw">c</span>(<span class="kw">min</span>(D[<span class="dv">2</span>,],<span class="dt">na.rm =</span> <span class="ot">TRUE</span>), <span class="kw">max</span>(D[<span class="dv">2</span>,],<span class="dt">na.rm =</span> <span class="ot">TRUE</span>)),<span class="dt">ylim=</span><span class="kw">c</span>(<span class="kw">min</span>(D[<span class="dv">1</span>,],<span class="dt">na.rm =</span> <span class="ot">TRUE</span>), <span class="kw">max</span>(D[<span class="dv">1</span>,],<span class="dt">na.rm =</span> <span class="ot">TRUE</span>)))
<span class="kw">points</span>(<span class="kw">t</span>(D[<span class="dv">1</span>,<span class="kw">which</span>(D[<span class="dv">3</span>,]<span class="op">==</span><span class="dv">1</span>)]),<span class="kw">t</span>(D[<span class="dv">2</span>,<span class="kw">which</span>(D[<span class="dv">3</span>,]<span class="op">==</span><span class="dv">1</span>)]),<span class="st">&#39;p&#39;</span>,<span class="dt">col=</span><span class="st">&#39;blue&#39;</span>)</code></pre></div>
<p><img src="04-dimensionality-reduction_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
<p>We can perform PCA in R using the  function. To do so, we must first filter out datapoints that have missing observations, as PCA does not, inherently, deal with missing observations:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">Dommitsamps &lt;-<span class="st"> </span><span class="kw">t</span>(<span class="kw">na.omit</span>(<span class="kw">t</span>(D[,]))); <span class="co">#Get the subset of samples</span>

pca1  &lt;-<span class="st"> </span><span class="kw">prcomp</span>(<span class="kw">t</span>(Dommitsamps[<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>,]), <span class="dt">center =</span> <span class="ot">TRUE</span>, <span class="dt">scale=</span><span class="ot">FALSE</span>)
ERexp &lt;-<span class="st"> </span>Dommitsamps[<span class="dv">3</span>,];

ER_neg &lt;-<span class="st"> </span>pca1<span class="op">$</span>x[<span class="kw">which</span>(ERexp<span class="op">==</span><span class="dv">0</span>),]
ER_pos &lt;-<span class="st"> </span>pca1<span class="op">$</span>x[<span class="kw">which</span>(ERexp<span class="op">==</span><span class="dv">1</span>),]

<span class="kw">plot</span>(ER_neg[,<span class="dv">1</span>],ER_neg[,<span class="dv">2</span>],<span class="st">&#39;p&#39;</span>,<span class="dt">col=</span><span class="st">&#39;red&#39;</span>, <span class="dt">xlab=</span><span class="st">&quot;PC1&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;PC2&quot;</span>,<span class="dt">xlim=</span><span class="kw">c</span>(<span class="op">-</span><span class="fl">4.5</span>, <span class="fl">4.2</span>),<span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">3</span>, <span class="fl">2.5</span>))
<span class="kw">points</span>(ER_pos[,<span class="dv">1</span>],ER_pos[,<span class="dv">2</span>],<span class="st">&#39;p&#39;</span>,<span class="dt">col=</span><span class="st">&#39;blue&#39;</span>)</code></pre></div>
<p><img src="04-dimensionality-reduction_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
<p>Note that the  has the option to centre and scale the data. That is, to normalise each variable to have a zero-mean and unit variance. This is particularly important when dealing with variables that may exist over very different scales. For example, for ecological datasets we may have variables that were measured in seconds with others measured in hours. Without normalisation there would appear to be much greater variance in the variable measured in seconds, potentially skewing the results. In general, when dealing with variables that are measured on similar scales (for example gene expression) it is not desirable to normalise the data.</p>
<p>We can better visualise what the PCA has done by plotting the original data side-by-side with the transformed data (note that here we have plotted the negative of PC1).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">plot</span>(<span class="kw">t</span>(D[<span class="dv">1</span>,<span class="kw">which</span>(D[<span class="dv">3</span>,]<span class="op">==</span><span class="dv">0</span>)]),<span class="kw">t</span>(D[<span class="dv">2</span>,<span class="kw">which</span>(D[<span class="dv">3</span>,]<span class="op">==</span><span class="dv">0</span>)]),<span class="st">&#39;p&#39;</span>,<span class="dt">col=</span><span class="st">&#39;red&#39;</span>, <span class="dt">ylab=</span><span class="st">&quot;XBP1&quot;</span>, <span class="dt">xlab=</span><span class="st">&quot;GATA3&quot;</span>,<span class="dt">xlim=</span><span class="kw">c</span>(<span class="kw">min</span>(D[<span class="dv">2</span>,],<span class="dt">na.rm =</span> <span class="ot">TRUE</span>), <span class="kw">max</span>(D[<span class="dv">2</span>,],<span class="dt">na.rm =</span> <span class="ot">TRUE</span>)),<span class="dt">ylim=</span><span class="kw">c</span>(<span class="kw">min</span>(D[<span class="dv">1</span>,],<span class="dt">na.rm =</span> <span class="ot">TRUE</span>), <span class="kw">max</span>(D[<span class="dv">1</span>,],<span class="dt">na.rm =</span> <span class="ot">TRUE</span>)))
<span class="kw">points</span>(<span class="kw">t</span>(D[<span class="dv">1</span>,<span class="kw">which</span>(D[<span class="dv">3</span>,]<span class="op">==</span><span class="dv">1</span>)]),<span class="kw">t</span>(D[<span class="dv">2</span>,<span class="kw">which</span>(D[<span class="dv">3</span>,]<span class="op">==</span><span class="dv">1</span>)]),<span class="st">&#39;p&#39;</span>,<span class="dt">col=</span><span class="st">&#39;blue&#39;</span>)
<span class="kw">plot</span>(<span class="op">-</span>ER_neg[,<span class="dv">1</span>],ER_neg[,<span class="dv">2</span>],<span class="st">&#39;p&#39;</span>,<span class="dt">col=</span><span class="st">&#39;red&#39;</span>, <span class="dt">xlab=</span><span class="st">&quot;-PC1&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;PC2&quot;</span>,<span class="dt">xlim=</span><span class="kw">c</span>(<span class="op">-</span><span class="fl">4.5</span>, <span class="fl">4.2</span>),<span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">3</span>, <span class="fl">2.5</span>))
<span class="kw">points</span>(<span class="op">-</span>ER_pos[,<span class="dv">1</span>],ER_pos[,<span class="dv">2</span>],<span class="st">&#39;p&#39;</span>,<span class="dt">col=</span><span class="st">&#39;blue&#39;</span>)</code></pre></div>
<p><img src="04-dimensionality-reduction_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<p>We can seen that we have simply rotated the original data, so that the greatest variance aligns along the x-axis and so forth. We can find out how much of the variance each of the principle components explains by looking at  variable:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))
<span class="kw">barplot</span>(((pca1<span class="op">$</span>sdev)<span class="op">^</span><span class="dv">2</span> <span class="op">/</span><span class="st"> </span><span class="kw">sum</span>(pca1<span class="op">$</span>sdev<span class="op">^</span><span class="dv">2</span>))<span class="op">*</span><span class="dv">100</span>, <span class="dt">names.arg=</span><span class="kw">c</span>(<span class="st">&quot;PC1&quot;</span>,<span class="st">&quot;PC2&quot;</span>), <span class="dt">ylab=</span><span class="st">&quot;% variance&quot;</span>)</code></pre></div>
<p><img src="04-dimensionality-reduction_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<p>Here we can see that PC1 explains the vast majority of the variance in the observations (for this example we should be able to see this by eye). The dimensionality reduction step of PCA occurs when we choose to discard the later PCs. Of course, by doing so we loose information about the system, but this may be an acceptable loss compared to the increased interpretability achieved by visualising the system in lower dimensions. In the example below, we follow from <span class="citation">(Ringnér <a href="#ref-ringner2008principal">2008</a>)</span>, and visualise the data using only PC1.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">1</span>))
<span class="kw">plot</span>(<span class="op">-</span>ER_neg[,<span class="dv">1</span>],<span class="kw">matrix</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>, <span class="kw">length</span>(ER_neg[,<span class="dv">1</span>])),<span class="st">&#39;p&#39;</span>,<span class="dt">col=</span><span class="st">&#39;red&#39;</span>, <span class="dt">xlab=</span><span class="st">&quot;PC1&quot;</span>,<span class="dt">xlim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">4</span>, <span class="dv">3</span>),<span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="fl">1.5</span>,<span class="fl">1.5</span>),<span class="dt">yaxt=</span><span class="st">&quot;n&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;&quot;</span>)
<span class="kw">points</span>(<span class="op">-</span>ER_pos[,<span class="dv">1</span>],<span class="kw">matrix</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>, <span class="kw">length</span>(ER_pos[,<span class="dv">1</span>])),<span class="st">&#39;p&#39;</span>,<span class="dt">col=</span><span class="st">&#39;blue&#39;</span>)
<span class="kw">points</span>(<span class="op">-</span>ER_neg[,<span class="dv">1</span>],<span class="kw">matrix</span>(<span class="dv">1</span>, <span class="dv">1</span>, <span class="kw">length</span>(ER_neg[,<span class="dv">1</span>])),<span class="st">&#39;p&#39;</span>,<span class="dt">col=</span><span class="st">&#39;red&#39;</span>, <span class="dt">xlab=</span><span class="st">&quot;PC1&quot;</span>,<span class="dt">xlim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">4</span>, <span class="dv">3</span>))
<span class="kw">points</span>(<span class="op">-</span>ER_pos[,<span class="dv">1</span>],<span class="kw">matrix</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="kw">length</span>(ER_pos[,<span class="dv">1</span>])),<span class="st">&#39;p&#39;</span>,<span class="dt">col=</span><span class="st">&#39;blue&#39;</span>)
<span class="kw">axis</span>(<span class="dt">side =</span> <span class="dv">2</span>, <span class="dt">at =</span> <span class="kw">seq</span>(<span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>, <span class="dt">by =</span> <span class="dv">1</span>), <span class="dt">labels =</span> <span class="kw">c</span>(<span class="st">&quot;All&quot;</span>,<span class="st">&quot;ER-&quot;</span>,<span class="st">&quot;ER+&quot;</span>))</code></pre></div>
<p><img src="04-dimensionality-reduction_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<p>So reducing the system down to one dimension appears to have done a good job at separating out the ER<span class="math inline">\(^+\)</span> cells from the ER<span class="math inline">\(^-\)</span> cells, suggesting that it may be of biological use. Precisely how many PCs to retain remains subjective. For visualisation purposed, it is typical to look at the first two or three only. However, when using PCA as an intermediate step within more complex workflows, more PCs are often retained e.g., by thresholding to a suitable level of explanatory variance.</p>
<div id="interpreting-the-principle-component-axes" class="section level3">
<h3><span class="header-section-number">5.1.1</span> Interpreting the Principle Component Axes</h3>
<p>In the original data, the individual axes had very obvious interpretations: the x-axis represented expression levels of <em>GATA3</em> and the y-axis represented the expression level of <em>XBP1</em>. Other than indicating maximum variance, what does PC1 mean? The individual axes represent linear combinations of the expression of various genes. This may not be immediately intuitive, but we can get a feel by projecting the original axes (gene expression) onto the (reduced dimensional) co-ordinate system.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genenames &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;GATA3&quot;</span>,<span class="st">&quot;XBP1&quot;</span>)
<span class="kw">plot</span>(<span class="op">-</span>pca1<span class="op">$</span>rotation[,<span class="dv">1</span>],pca1<span class="op">$</span>rotation[,<span class="dv">2</span>], <span class="dt">type=</span><span class="st">&quot;n&quot;</span>, <span class="dt">xlim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>, <span class="dv">2</span>), <span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">2</span>, <span class="dv">2</span>), <span class="dt">xlab=</span><span class="st">&quot;PC1&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;PC2&quot;</span>)
<span class="kw">text</span>(<span class="op">-</span>pca1<span class="op">$</span>rotation[,<span class="dv">1</span>], pca1<span class="op">$</span>rotation[,<span class="dv">2</span>], genenames, <span class="dt">cex =</span> .<span class="dv">4</span>)
<span class="kw">arrows</span>(<span class="dv">0</span>, <span class="dv">0</span>, <span class="dt">x1 =</span> <span class="op">-</span>pca1<span class="op">$</span>rotation[,<span class="dv">1</span>], <span class="dt">y1 =</span> <span class="op">-</span>pca1<span class="op">$</span>rotation[,<span class="dv">2</span>],<span class="dt">length=</span><span class="fl">0.1</span>)</code></pre></div>
<p><img src="04-dimensionality-reduction_files/figure-html/unnamed-chunk-7-1.png" width="672" /></p>
<p>In this particular case, we can see that both genes appear to be reasonably strongly associated with PC1. When dealing with much larger systems e.g., with more genes, we can, of course, project the original axes into the reduced dimensional space. In general this is particularly useful for identifying genes associated with particular PCs, and ultimately assigning a biological interpretation to the PCs.</p>
</div>
<div id="horseshoe-effect" class="section level3">
<h3><span class="header-section-number">5.1.2</span> Horseshoe effect</h3>
<p>Principle component analysis is a linear dimensionality reduction technique, and is not always appropriate for complex datasets, particularly when dealing with nonlinearities. To illustrate this, let’s consider an simulated expression set containing <span class="math inline">\(8\)</span> genes, with <span class="math inline">\(10\)</span> timepoints/conditions. We can represent this dataset in terms of a matrix:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">X &lt;-<span class="st"> </span><span class="kw">matrix</span>( <span class="kw">c</span>(<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">2</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,
                 <span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">2</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,
                 <span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">2</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,  
                 <span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">2</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,   
                 <span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">2</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,    
                 <span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">2</span>,<span class="dv">0</span>,<span class="dv">0</span>,   
                 <span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">2</span>,<span class="dv">0</span>,  
                 <span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">2</span>,<span class="dv">4</span>,<span class="dv">2</span>), <span class="dt">nrow=</span><span class="dv">8</span>,  <span class="dt">ncol=</span><span class="dv">10</span>, <span class="dt">byrow =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>Or we can visualise by plotting a few of the genes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plot</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,X[<span class="dv">1</span>,],<span class="dt">type=</span><span class="st">&quot;l&quot;</span>,<span class="dt">col=</span><span class="st">&quot;red&quot;</span>,<span class="dt">xlim=</span><span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">14</span>),<span class="dt">xlab=</span><span class="st">&quot;Time&quot;</span>,<span class="dt">ylab=</span><span class="st">&quot;Expression&quot;</span>)
<span class="kw">points</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,X[<span class="dv">2</span>,],<span class="dt">type=</span><span class="st">&quot;l&quot;</span>,<span class="dt">col=</span><span class="st">&quot;blue&quot;</span>)
<span class="kw">points</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>,X[<span class="dv">5</span>,],<span class="dt">type=</span><span class="st">&quot;l&quot;</span>,<span class="dt">col=</span><span class="st">&quot;black&quot;</span>)
<span class="kw">legend</span>(<span class="dv">8</span>, <span class="dv">4</span>, <span class="dt">legend=</span><span class="kw">c</span>(<span class="st">&quot;gene 1&quot;</span>, <span class="st">&quot;gene 2&quot;</span>, <span class="st">&quot;gene 5&quot;</span>), <span class="dt">col=</span><span class="kw">c</span>(<span class="st">&quot;red&quot;</span>, <span class="st">&quot;blue&quot;</span>, <span class="st">&quot;black&quot;</span>),<span class="dt">lty=</span><span class="dv">1</span>, <span class="dt">cex=</span><span class="fl">0.8</span>)</code></pre></div>
<p><img src="04-dimensionality-reduction_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
<p>By eye, we see that the data can be separated out by a single direction: that is, we can order the data from time/condition 1 through to time/condition 10. Intuitively, then, the data can be represented by a single dimension. Let’s run PCA as we would normally, and visualise the result, plotting the first two PCs:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pca2 &lt;-<span class="st"> </span><span class="kw">prcomp</span>(<span class="kw">t</span>(X),<span class="dt">center =</span> <span class="ot">TRUE</span>,<span class="dt">scale=</span><span class="ot">FALSE</span>)
condnames =<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;TP1&#39;</span>,<span class="st">&#39;TP2&#39;</span>,<span class="st">&#39;TP3&#39;</span>,<span class="st">&#39;TP4&#39;</span>,<span class="st">&#39;TP5&#39;</span>,<span class="st">&#39;TP6&#39;</span>,<span class="st">&#39;TP7&#39;</span>,<span class="st">&#39;TP8&#39;</span>,<span class="st">&#39;TP9&#39;</span>,<span class="st">&#39;TP10&#39;</span>)
<span class="kw">plot</span>(pca2<span class="op">$</span>x[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>],<span class="dt">type=</span><span class="st">&quot;p&quot;</span>,<span class="dt">col=</span><span class="st">&quot;red&quot;</span>,<span class="dt">xlim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">5</span>, <span class="dv">5</span>),<span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">5</span>, <span class="dv">5</span>))
<span class="kw">text</span>(pca2<span class="op">$</span>x[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]<span class="op">+</span><span class="fl">0.5</span>, condnames, <span class="dt">cex =</span> <span class="fl">0.7</span>)</code></pre></div>
<p><img src="04-dimensionality-reduction_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
<p>We see that the PCA plot has placed the datapoints in a horseshoe shape, with condition/time point 1 very close to condition/time point 10. From the earlier plots of gene expression profiles we can see that the relationships between the various genes are not entirely straightforward. For example, gene 1 is initially correlated with gene 2, then negatively correlated, and finally uncorrelated, whilst no correlation exists between gene 1 and genes 5 - 8. These nonlinearities make it difficult for PCA which, in general, attempts to preserve large pairwise distances, leading to the well known horseshoe effect <span class="citation">(Novembre and Stephens <a href="#ref-novembre2008interpreting">2008</a>,<span class="citation">Reich, Price, and Patterson (<a href="#ref-reich2008principal">2008</a>)</span>)</span>. These types of artefacts may be problematic when trying to interpret data, and due care must be given when these type of effects are seen.</p>
</div>
<div id="pca-analysis-of-mammalian-development" class="section level3">
<h3><span class="header-section-number">5.1.3</span> PCA analysis of mammalian development</h3>
<p>Now that we have a feel for PCA and understand some of the basic commands we can apply it in a real setting. Here we will make use of preprocessed data taken from <span class="citation">(L. Yan et al. <a href="#ref-yan2013single">2013</a>)</span> (GEO GSE36552) and <span class="citation">(F. Guo et al. <a href="#ref-guo2015transcriptome">2015</a>)</span> (GEO GSE63818). The data from <span class="citation">(L. Yan et al. <a href="#ref-yan2013single">2013</a>)</span> represents single cell RNA-seq measurements from human embryos from the zygote stage (a single cell produced following fertilisation of an egg) through to the blastocyst stage (an embryo consisting of around 64 cells), as well as human embryonic stem cells (hESC; cells extracted from an early blsatocyst stage embryo and maintained <em>in vitro</em>). The dataset of <span class="citation">(F. Guo et al. <a href="#ref-guo2015transcriptome">2015</a>)</span> contains scRNA-seq data from human primordial germ cells (hPGCs), precursors of sperm or eggs that are specified early in the developing human embryo soon after implantation (around week 2-3 in humans), and somatic cells. Together, these datasets provide useful insights into early human development, and possible mechanisms for the specification of early cell types, such as PGCs.</p>
<p>Preprocessed data contains <span class="math inline">\(\log_2\)</span> normalised counts for around <span class="math inline">\(400\)</span> cells using <span class="math inline">\(2957\)</span> marker genes can be found in the file . Note that the first line of data in the file is an indicator denoting cell type (-1 = ESC, 0 = pre-implantation, 1 = PGC, and 2 = somatic cell). The second row indicates the sex of the cell (0 = unknown/unlabelled, 1 = XX, 2 = XY), with the third row indicating capture time (-1 = ESC, 0 - 7 denotes various developmental stages from zygote to blastocyst, 8 - 13 indicates increasing times of embryo development from week 4 through to week 19).</p>
<p>Exercise 5.1. First load in the expression data into R and plot some example expression patterns.</p>
<p>Exercise 5.2. Use  to perform PCA on the data.</p>
<p>Exercise 5.3. Try plotting visualising the original axis. Can we identify any genes of interest that may be particularly important for PGCs?</p>
<p>Exercise 5.4. Does the data separate well? Perform k-means cluster analysis on the data to see if we can identify distinct clusters.</p>
<p>Exercise 5.5. Perform a differential expression analysis between blastocyst cells and the PGCs.</p>
</div>
</div>
<div id="nonlinear-dimensionality-reduction" class="section level2">
<h2><span class="header-section-number">5.2</span> Nonlinear Dimensionality Reduction</h2>
<p>Whilst <span id="linear-dimensionality-reduction">PCA</span> is extremely useful for exploratory analysis, it is not always appropriate, particularly for datasets with nonlinearities. A large number of nonlinear dimensionality reduction techniques have therefore been developed. Perhaps the most commonly applied technique is t-distributed stochastic neighbour embedding (tSNE) <span class="citation">(L. van der Maaten and Hinton <a href="#ref-maaten2008visualizing">2008</a>,<span class="citation">L. van der Maaten (<a href="#ref-van2009learning">2009</a>)</span>,<span class="citation">Van der Maaten and Hinton (<a href="#ref-van2012visualizing">2012</a>)</span>,<span class="citation">Van Der Maaten (<a href="#ref-van2014accelerating">2014</a>)</span>)</span>.</p>
<p>In general, tSNE attempts to take points in a high-dimensional space and find a faithful representation of those points in a lower-dimensional space. The SNE algorithm initially converts the high-dimensional Euclidean distances between datapoints into conditional probabilities. Here <span class="math inline">\(p_{j|i}\)</span>, indicates the probability that datapoint <span class="math inline">\(x_i\)</span> would pick <span class="math inline">\(x_j\)</span> as its neighbour if neighbours were picked in proportion to their probability density under a Gaussian centred at <span class="math inline">\(x_i\)</span>:</p>
<p><span class="math inline">\(p_{j|i} = \frac{\exp(-|\mathbf{x}_i - \mathbf{x}_j|^2/2\sigma_i^2)}{\sum_{k\neq l}\exp(-|\mathbf{x}_k - \mathbf{x}_l|^2/2\sigma_i^2)}\)</span></p>
<p>We can define a similar conditional probability for the datapoints in the reduced dimensional space, <span class="math inline">\(y_j\)</span> and <span class="math inline">\(y_j\)</span> as:</p>
<p><span class="math inline">\(q_{j|i} = \frac{\exp(-|\mathbf{y}_i - \mathbf{y}_j|^2)}{\sum_{k\neq l}\exp(-|\mathbf{y}_k - \mathbf{y}_l|^2)}\)</span>.</p>
<p>Natural extensions to this would instead use a Student-t distribution for the lower dimensional space.</p>
<p><span class="math inline">\(q_{j|i} = \frac{(1+|\mathbf{y}_i - \mathbf{y}_j|^2)^{-1}}{\sum_{k\neq l}(1+|\mathbf{y}_i - \mathbf{y}_j|^2)^{-1}}\)</span>.</p>
<p>If SNE has mapped points <span class="math inline">\(\mathbf{y}_i\)</span> and <span class="math inline">\(\mathbf{y}_j\)</span> faithfully, we have <span class="math inline">\(p_{j|i} = q_{j|i}\)</span>. We can define a similarity measure over these distribution based on the Kullback-Leibler-divergence:</p>
<p><span class="math inline">\(C = \sum KL(P_i||Q_i)= \sum_i \sum_j p_{i|j} \log \biggl{(} \frac{p_{i|j}}{q_{i|j}} \biggr{)}\)</span></p>
<p>If <span class="math inline">\(p_{j|i} = q_{j|i}\)</span>, that is, if our reduced dimensionality representation faithfully captures the higher dimensional data, this value will be equal to zero, otherwise it will be a positive number. We can attempt to minimise this value using gradient descent.</p>
<p>Note that in many cases this lower dimensionality space can be initialised using PCA or other dimensionality reduction technique. The tSNE algorithm is implemented in R via the  package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(Rtsne)
<span class="kw">library</span>(scatterplot3d)
<span class="kw">set.seed</span>(<span class="dv">12345</span>)</code></pre></div>
<p>To get a feel for tSNE we will first generate some artificial data. In this case we generate two different groups that exist in a 3-dimensional space. We choose these groups to be Gaussian distributed, with different means and variances:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">D1 &lt;-<span class="st"> </span><span class="kw">matrix</span>( <span class="kw">rnorm</span>(<span class="dv">5</span><span class="op">*</span><span class="dv">3</span>,<span class="dt">mean=</span><span class="dv">0</span>,<span class="dt">sd=</span><span class="dv">1</span>), <span class="dv">100</span>, <span class="dv">3</span>) 
D2 &lt;-<span class="st"> </span><span class="kw">matrix</span>( <span class="kw">rnorm</span>(<span class="dv">5</span><span class="op">*</span><span class="dv">3</span>,<span class="dt">mean=</span><span class="dv">5</span>,<span class="dt">sd=</span><span class="dv">3</span>), <span class="dv">100</span>, <span class="dv">3</span>) 
G1 &lt;-<span class="st"> </span><span class="kw">matrix</span>( <span class="dv">1</span>, <span class="dv">100</span>, <span class="dv">1</span>) 
G2 &lt;-<span class="st"> </span><span class="kw">matrix</span>( <span class="dv">2</span>, <span class="dv">100</span>, <span class="dv">1</span>) 
D3 &lt;-<span class="st"> </span><span class="kw">rbind</span>(D1,D2)
G3 &lt;-<span class="st"> </span><span class="kw">rbind</span>(G1,G2)
colors &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;red&quot;</span>, <span class="st">&quot;blue&quot;</span>)
colors &lt;-<span class="st"> </span>colors[G3]
<span class="kw">scatterplot3d</span>(D3,<span class="dt">color=</span>colors, <span class="dt">main=</span><span class="st">&quot;3D Scatterplot&quot;</span>,<span class="dt">xlab=</span><span class="st">&quot;x&quot;</span>,<span class="dt">ylab=</span><span class="st">&quot;y&quot;</span>,<span class="dt">zlab=</span><span class="st">&quot;z&quot;</span>)</code></pre></div>
<p><img src="04-dimensionality-reduction_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
<p>We can run tSNE on this dataset and try to condense the data down from a three-dimensional to a two-dimensional representation. Unlike PCA, which has no real free parameters, tSNE has a variety of parameters that need to be set. First, we have the perplexity parameter which, in essence, balances local and global aspects of the data. For low values of perplexity, the algorithm will tend to entirely focus on keeping datapoints locally together.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">tsne_model_<span class="dv">1</span> &lt;-<span class="st"> </span><span class="kw">Rtsne</span>(<span class="kw">as.matrix</span>(D3), <span class="dt">check_duplicates=</span><span class="ot">FALSE</span>, <span class="dt">pca=</span><span class="ot">TRUE</span>, <span class="dt">perplexity=</span><span class="dv">10</span>, <span class="dt">theta=</span><span class="fl">0.5</span>, <span class="dt">dims=</span><span class="dv">2</span>)
y1 &lt;-<span class="st"> </span>tsne_model_<span class="dv">1</span><span class="op">$</span>Y[<span class="kw">which</span>(D[<span class="dv">1</span>,]<span class="op">==-</span><span class="dv">1</span>),<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]
tsne_model_<span class="dv">1</span> &lt;-<span class="st"> </span><span class="kw">Rtsne</span>(<span class="kw">as.matrix</span>(D3), <span class="dt">check_duplicates=</span><span class="ot">FALSE</span>, <span class="dt">pca=</span><span class="ot">TRUE</span>, <span class="dt">perplexity=</span><span class="dv">10</span>, <span class="dt">theta=</span><span class="fl">0.5</span>, <span class="dt">dims=</span><span class="dv">2</span>)

<span class="kw">plot</span>(tsne_model_<span class="dv">1</span><span class="op">$</span>Y[<span class="dv">1</span><span class="op">:</span><span class="dv">100</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>],<span class="dt">type=</span><span class="st">&quot;p&quot;</span>,<span class="dt">col=</span><span class="st">&quot;red&quot;</span>,<span class="dt">xlim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">45</span>, <span class="dv">45</span>),<span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">45</span>, <span class="dv">45</span>),<span class="dt">xlab=</span><span class="st">&quot;tSNE1&quot;</span>,<span class="dt">ylab=</span><span class="st">&quot;tSNE1&quot;</span>)
<span class="kw">points</span>(tsne_model_<span class="dv">1</span><span class="op">$</span>Y[<span class="dv">101</span><span class="op">:</span><span class="dv">200</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>],<span class="dt">type=</span><span class="st">&quot;p&quot;</span>,<span class="dt">col=</span><span class="st">&quot;blue&quot;</span>)</code></pre></div>
<p><img src="04-dimensionality-reduction_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
<p>Note that here we have set the perplexity parameter reasonably low, and tSNE appears to have identified a lot of local structure that (we know) doesn’t exist. Let’s try again using a larger value for the perplexity parameter.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">y1 &lt;-<span class="st"> </span>tsne_model_<span class="dv">1</span><span class="op">$</span>Y[<span class="kw">which</span>(D[<span class="dv">1</span>,]<span class="op">==-</span><span class="dv">1</span>),<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>]
tsne_model_<span class="dv">1</span> &lt;-<span class="st"> </span><span class="kw">Rtsne</span>(<span class="kw">as.matrix</span>(D3), <span class="dt">check_duplicates=</span><span class="ot">FALSE</span>, <span class="dt">pca=</span><span class="ot">TRUE</span>, <span class="dt">perplexity=</span><span class="dv">50</span>, <span class="dt">theta=</span><span class="fl">0.5</span>, <span class="dt">dims=</span><span class="dv">2</span>)

<span class="kw">plot</span>(tsne_model_<span class="dv">1</span><span class="op">$</span>Y[<span class="dv">1</span><span class="op">:</span><span class="dv">100</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>],<span class="dt">type=</span><span class="st">&quot;p&quot;</span>,<span class="dt">col=</span><span class="st">&quot;red&quot;</span>,<span class="dt">xlim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">45</span>, <span class="dv">45</span>),<span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">45</span>, <span class="dv">45</span>),<span class="dt">xlab=</span><span class="st">&quot;tSNE1&quot;</span>,<span class="dt">ylab=</span><span class="st">&quot;tSNE2&quot;</span>)
<span class="kw">points</span>(tsne_model_<span class="dv">1</span><span class="op">$</span>Y[<span class="dv">101</span><span class="op">:</span><span class="dv">200</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>],<span class="dt">type=</span><span class="st">&quot;p&quot;</span>,<span class="dt">col=</span><span class="st">&quot;blue&quot;</span>)</code></pre></div>
<p><img src="04-dimensionality-reduction_files/figure-html/unnamed-chunk-14-1.png" width="672" /></p>
<p>This appears to have done a better job of representing the data in a two-dimensional space.</p>
<div id="nonlinear-warping" class="section level3">
<h3><span class="header-section-number">5.2.1</span> Nonlinear warping</h3>
<p>In our previous example we showed that if the perplexity parameter was correctly set, tSNE seperated out the two populations very well. If we plot the original data next to the tSNE reduced dimensionality represention, however, we will notice something interesting:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">scatterplot3d</span>(D3,<span class="dt">color=</span>colors, <span class="dt">main=</span><span class="st">&quot;3D Scatterplot&quot;</span>,<span class="dt">xlab=</span><span class="st">&quot;x&quot;</span>,<span class="dt">ylab=</span><span class="st">&quot;y&quot;</span>,<span class="dt">zlab=</span><span class="st">&quot;z&quot;</span>)
<span class="kw">plot</span>(tsne_model_<span class="dv">1</span><span class="op">$</span>Y[<span class="dv">1</span><span class="op">:</span><span class="dv">100</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>],<span class="dt">type=</span><span class="st">&quot;p&quot;</span>,<span class="dt">col=</span><span class="st">&quot;red&quot;</span>,<span class="dt">xlim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">45</span>, <span class="dv">45</span>),<span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">45</span>, <span class="dv">45</span>),<span class="dt">xlab=</span><span class="st">&quot;tSNE1&quot;</span>, <span class="dt">ylab=</span><span class="st">&quot;tSNE2&quot;</span>)
<span class="kw">points</span>(tsne_model_<span class="dv">1</span><span class="op">$</span>Y[<span class="dv">101</span><span class="op">:</span><span class="dv">200</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>],<span class="dt">type=</span><span class="st">&quot;p&quot;</span>,<span class="dt">col=</span><span class="st">&quot;blue&quot;</span>)</code></pre></div>
<p><img src="04-dimensionality-reduction_files/figure-html/unnamed-chunk-15-1.png" width="672" /></p>
<p>Whilst in the origianl data the two groups had very different variances, in the reduced dimensionality representation they appeared to show a similar spread. This is down to tSNEs ability to represent nonlinearities, and the algorithm performs different transformations on different regions. This is, of course, important to keep in mind: the spread in a tSNE output are not always indicative of the level of heterogeneity in the data.</p>
</div>
<div id="stochasticity" class="section level3">
<h3><span class="header-section-number">5.2.2</span> Stochasticity</h3>
<p>A final important point to note is that tSNE is stochastic in nature. Unlike PCA which, for the same dataset, will always yield the same result, if you run tSNE twice you will likely find different results. We can illustrate this below, by running tSNE again for perplexity <span class="math inline">\(30\)</span>, and plotting the results alongside the previous ones.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">set.seed</span>(<span class="dv">123456</span>)

tsne_model_<span class="dv">1</span> &lt;-<span class="st"> </span><span class="kw">Rtsne</span>(<span class="kw">as.matrix</span>(D3), <span class="dt">check_duplicates=</span><span class="ot">FALSE</span>, <span class="dt">pca=</span><span class="ot">TRUE</span>, <span class="dt">perplexity=</span><span class="dv">30</span>, <span class="dt">theta=</span><span class="fl">0.5</span>, <span class="dt">dims=</span><span class="dv">2</span>)

tsne_model_<span class="dv">2</span> &lt;-<span class="st"> </span><span class="kw">Rtsne</span>(<span class="kw">as.matrix</span>(D3), <span class="dt">check_duplicates=</span><span class="ot">FALSE</span>, <span class="dt">pca=</span><span class="ot">TRUE</span>, <span class="dt">perplexity=</span><span class="dv">30</span>, <span class="dt">theta=</span><span class="fl">0.5</span>, <span class="dt">dims=</span><span class="dv">2</span>)

<span class="kw">par</span>(<span class="dt">mfrow=</span><span class="kw">c</span>(<span class="dv">1</span>,<span class="dv">2</span>))
<span class="kw">plot</span>(tsne_model_<span class="dv">1</span><span class="op">$</span>Y[<span class="dv">1</span><span class="op">:</span><span class="dv">100</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>],<span class="dt">type=</span><span class="st">&quot;p&quot;</span>,<span class="dt">col=</span><span class="st">&quot;red&quot;</span>,<span class="dt">xlim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">45</span>, <span class="dv">45</span>),<span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">45</span>, <span class="dv">45</span>),<span class="dt">xlab=</span><span class="st">&quot;tSNE1&quot;</span>,<span class="dt">ylab=</span><span class="st">&quot;tSNE2&quot;</span>)
<span class="kw">points</span>(tsne_model_<span class="dv">1</span><span class="op">$</span>Y[<span class="dv">101</span><span class="op">:</span><span class="dv">200</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>],<span class="dt">type=</span><span class="st">&quot;p&quot;</span>,<span class="dt">col=</span><span class="st">&quot;blue&quot;</span>)

<span class="kw">plot</span>(tsne_model_<span class="dv">2</span><span class="op">$</span>Y[<span class="dv">1</span><span class="op">:</span><span class="dv">100</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>],<span class="dt">type=</span><span class="st">&quot;p&quot;</span>,<span class="dt">col=</span><span class="st">&quot;red&quot;</span>,<span class="dt">xlim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">45</span>, <span class="dv">45</span>),<span class="dt">ylim=</span><span class="kw">c</span>(<span class="op">-</span><span class="dv">45</span>, <span class="dv">45</span>),<span class="dt">xlab=</span><span class="st">&quot;tSNE1&quot;</span>,<span class="dt">ylab=</span><span class="st">&quot;tSNE2&quot;</span>)
<span class="kw">points</span>(tsne_model_<span class="dv">2</span><span class="op">$</span>Y[<span class="dv">101</span><span class="op">:</span><span class="dv">200</span>,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>],<span class="dt">type=</span><span class="st">&quot;p&quot;</span>,<span class="dt">col=</span><span class="st">&quot;blue&quot;</span>)</code></pre></div>
<p><img src="04-dimensionality-reduction_files/figure-html/unnamed-chunk-16-1.png" width="672" /></p>
<p>Note that this stochasticity, itself, may be a useful property, allowing us to gauge robustness of our biological interpretations. A comprehensive blog discussing the various pitfalls of tSNE is available <a href="https://distill.pub/2016/misread-tsne/">here</a>.</p>
</div>
<div id="analysis-of-mammalian-development" class="section level3">
<h3><span class="header-section-number">5.2.3</span> Analysis of mammalian development</h3>
<p>In earlier sections we used PCA to analyse scRNA-seq datasets of early human embryo development. In general PCA seemed adept at picking out different cell types and idetifying putative regulators associated with those cell types. We will now use tSNE to analyse the same data.</p>
<p>Excercise 8.6. Load in the single cell dataset and run tSNE.</p>
<p>Excercise 8.7.Note that cells labelled as pre-implantation actually consists of a variety of cells, from oocytes through to blastocyst stage. Take a look at the pre-implantation cells only using tSNE. Hint: a more refined categorisation of the developmental stage of pre-implantation cells can be found by looking at the developmental time variable (0=oocyte, 1=zygote, 2=2C, 3=4C, 4=8C, 5=Morula, 6=blastocyst). Try plotting the data from tSNE colouring the data according to developmental stage.</p>
<p>Excercise 8.8. How do pre-implantation cells cluster in tSNE?</p>
</div>
</div>
<div id="other-dimensionality-reduction-techniques" class="section level2">
<h2><span class="header-section-number">5.3</span> Other dimensionality reduction techniques</h2>
<p>A large number of alternative dimensionality reduction techniques exist with corresponding implementation in R. These include probabilistic extensions to PCA <a href="https://www.rdocumentation.org/packages/pcaMethods/versions/1.64.0">pcaMethods</a>, as well as other nonlinear dimensionality reduction techniques <a href="https://www.rdocumentation.org/packages/RDRToolbox/versions/1.22.0">Isomap</a>, as well as those based on Gaussian Processes (<a href="https://github.com/SheffieldML/vargplvm.git">GPLVM</a>; Lawrence 2004). Other packages such as <a href="https://cran.r-project.org/web/packages/kernlab/index.html">kernlab</a> provide a general suite of tools for dimensionality reduction.</p>
<p>Solutions to exercises can be found in appendix <a href="solutions-dimensionality-reduction.html#solutions-dimensionality-reduction">D</a>.</p>

</div>
</div>
<h3>References</h3>
<div id="refs" class="references">
<div id="ref-Breeze873">
<p>Breeze, Emily, Elizabeth Harrison, Stuart McHattie, Linda Hughes, Richard Hickman, Claire Hill, Steven Kiddle, et al. 2011. “High-Resolution Temporal Profiling of Transcripts During Arabidopsis Leaf Senescence Reveals a Distinct Chronology of Processes and Regulation.” <em>The Plant Cell Online</em> 23 (3). American Society of Plant Biologists: 873–94. doi:<a href="https://doi.org/10.1105/tpc.111.083345">10.1105/tpc.111.083345</a>.</p>
</div>
<div id="ref-windram2012arabidopsis">
<p>Windram, Oliver, Priyadharshini Madhou, Stuart McHattie, Claire Hill, Richard Hickman, Emma Cooke, Dafyd J Jenkins, et al. 2012. “Arabidopsis Defense Against Botrytis Cinerea: Chronology and Regulation Deciphered by High-Resolution Temporal Transcriptomic Analysis.” <em>The Plant Cell</em> 24 (9). Am Soc Plant Biol: 3530–57.</p>
</div>
<div id="ref-Lewis15">
<p>Lewis, Laura A, Krzysztof Polanski, Marta de Torres-Zabala, Siddharth Jayaraman, Laura Bowden, Jonathan Moore, Christopher A, et al. 2015. “Transcriptional Dynamics Driving MAMP-Triggered Immunity and Pathogen Effector-Mediated Immunosuppression in Arabidopsis Leaves Following Infection with Pseudomonas Syringae Pv Tomato Dc3000.” <em>The Plant Cell</em> 27 (11). Am Soc Plant Biol: 3038–64.</p>
</div>
<div id="ref-Bechtold">
<p>Bechtold, Ulrike, Christopher A Penfold, Dafyd J Jenkins, Roxane Legaie, Jonathan D Moore, Tracy Lawson, Jack SA Matthews, et al. 2016. “Time-Series Transcriptomics Reveals That Agamous-Like22 Links Primary Metabolism to Developmental Processes in Drought-Stressed Arabidopsis.” <em>The Plant Cell Online</em>. American Society of Plant Biologists. doi:<a href="https://doi.org/10.1105/tpc.15.00910">10.1105/tpc.15.00910</a>.</p>
</div>
<div id="ref-irie2015sox17">
<p>Irie, Naoko, Leehee Weinberger, Walfred WC Tang, Toshihiro Kobayashi, Sergey Viukov, Yair S Manor, Sabine Dietmann, Jacob H Hanna, and M Azim Surani. 2015. “SOX17 Is a Critical Specifier of Human Primordial Germ Cell Fate.” <em>Cell</em> 160 (1). Elsevier: 253–68.</p>
</div>
<div id="ref-tang2015unique">
<p>Tang, Walfred WC, Sabine Dietmann, Naoko Irie, Harry G Leitch, Vasileios I Floros, Charles R Bradshaw, Jamie A Hackett, Patrick F Chinnery, and M Azim Surani. 2015. “A Unique Gene Regulatory Network Resets the Human Germline Epigenome for Development.” <em>Cell</em> 161 (6). Elsevier: 1453–67.</p>
</div>
<div id="ref-tang2009mrna">
<p>Tang, Fuchou, Catalin Barbacioru, Yangzhou Wang, Ellen Nordman, Clarence Lee, Nanlan Xu, Xiaohui Wang, et al. 2009. “MRNA-Seq Whole-Transcriptome Analysis of a Single Cell.” <em>Nature Methods</em> 6 (5). Nature Publishing Group: 377–82.</p>
</div>
<div id="ref-svensson2017moore">
<p>Svensson, Valentine, Roser Vento-Tormo, and Sarah A Teichmann. 2017. “Moore’s Law in Single Cell Transcriptomics.” <em>arXiv Preprint arXiv:1704.01379</em>.</p>
</div>
<div id="ref-pearson1901liii">
<p>Pearson, Karl. 1901. “LIII. on Lines and Planes of Closest Fit to Systems of Points in Space.” <em>The London, Edinburgh, and Dublin Philosophical Magazine and Journal of Science</em> 2 (11). Taylor &amp; Francis: 559–72.</p>
</div>
<div id="ref-hotelling1933analysis">
<p>Hotelling, Harold. 1933. “Analysis of a Complex of Statistical Variables into Principal Components.” <em>Journal of Educational Psychology</em> 24 (6). Warwick &amp; York: 417.</p>
</div>
<div id="ref-sforza1964analysis">
<p>Sforza, Cavalli LL, and Anthony William Fairbank Edwards. 1964. “Analysis of Human Evolution.” <em>Genet. Today</em> 3: 923–33.</p>
</div>
<div id="ref-vohradsky1997identification">
<p>Vohradsky, Jiří, Xin-Ming Li, and Charles J Thompson. 1997. “Identification of Procaryotic Developmental Stages by Statistical Analyses of Two-Dimensional Gel Patterns.” <em>Electrophoresis</em> 18 (8). Wiley Online Library: 1418–28.</p>
</div>
<div id="ref-craig1997developmental">
<p>Craig, JC, JH Eberwine, JA Calvin, B Wlodarczyk, GD Bennett, and RH Finnell. 1997. “Developmental Expression of Morphoregulatory Genes in the Mouse Embryo: An Analytical Approach Using a Novel Technology.” <em>Biochemical and Molecular Medicine</em> 60 (2). Elsevier: 81–91.</p>
</div>
<div id="ref-hilsenbeck1999statistical">
<p>Hilsenbeck, Susan G, William E Friedrichs, Rachel Schiff, Peter O’connell, Rhonda K Hansen, C Kent Osborne, and Suzanne AW Fuqua. 1999. “Statistical Analysis of Array Expression Data as Applied to the Problem of Tamoxifen Resistance.” <em>Journal of the National Cancer Institute</em> 91 (5). Oxford University Press: 453–59.</p>
</div>
<div id="ref-ringner2008principal">
<p>Ringnér, Markus. 2008. “What Is Principal Component Analysis?” <em>Nature Biotechnology</em> 26 (3). Nature Publishing Group: 303–4.</p>
</div>
<div id="ref-saal2007poor">
<p>Saal, Lao H, Peter Johansson, Karolina Holm, Sofia K Gruvberger-Saal, Qing-Bai She, Matthew Maurer, Susan Koujak, et al. 2007. “Poor Prognosis in Carcinoma Is Associated with a Gene Expression Signature of Aberrant Pten Tumor Suppressor Pathway Activity.” <em>Proceedings of the National Academy of Sciences</em> 104 (18). National Acad Sciences: 7564–9.</p>
</div>
<div id="ref-novembre2008interpreting">
<p>Novembre, John, and Matthew Stephens. 2008. “Interpreting Principal Component Analyses of Spatial Population Genetic Variation.” <em>Nature Genetics</em> 40 (5). Nature Publishing Group: 646–49.</p>
</div>
<div id="ref-reich2008principal">
<p>Reich, David, Alkes L Price, and Nick Patterson. 2008. “Principal Component Analysis of Genetic Data.” <em>Nature Genetics</em> 40 (5). Nature Publishing Group: 491–92.</p>
</div>
<div id="ref-yan2013single">
<p>Yan, Liying, Mingyu Yang, Hongshan Guo, Lu Yang, Jun Wu, Rong Li, Ping Liu, et al. 2013. “Single-Cell RNA-Seq Profiling of Human Preimplantation Embryos and Embryonic Stem Cells.” <em>Nature Structural &amp; Molecular Biology</em> 20 (9). Nature Publishing Group: 1131–9.</p>
</div>
<div id="ref-guo2015transcriptome">
<p>Guo, Fan, Liying Yan, Hongshan Guo, Lin Li, Boqiang Hu, Yangyu Zhao, Jun Yong, et al. 2015. “The Transcriptome and DNA Methylome Landscapes of Human Primordial Germ Cells.” <em>Cell</em> 161 (6). Elsevier: 1437–52.</p>
</div>
<div id="ref-maaten2008visualizing">
<p>Maaten, Laurens van der, and Geoffrey Hinton. 2008. “Visualizing Data Using T-Sne.” <em>Journal of Machine Learning Research</em> 9 (Nov): 2579–2605.</p>
</div>
<div id="ref-van2009learning">
<p>Maaten, Laurens van der. 2009. “Learning a Parametric Embedding by Preserving Local Structure.” <em>RBM</em> 500 (500): 26.</p>
</div>
<div id="ref-van2012visualizing">
<p>Van der Maaten, Laurens, and Geoffrey Hinton. 2012. “Visualizing Non-Metric Similarities in Multiple Maps.” <em>Machine Learning</em> 87 (1). Springer: 33–55.</p>
</div>
<div id="ref-van2014accelerating">
<p>Van Der Maaten, Laurens. 2014. “Accelerating T-Sne Using Tree-Based Algorithms.” <em>Journal of Machine Learning Research</em> 15 (1): 3221–45.</p>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="logistic-regression.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="clustering.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"google": false,
"linkedin": false,
"weibo": false,
"instapper": false,
"vk": false,
"all": ["facebook", "google", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/bioinformatics-training/intro-machine-learning/edit/master/04-dimensionality-reduction.Rmd",
"text": "Edit"
},
"download": ["intro-machine-learning.pdf"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "";
    if (src === "" || src === "true") src = "https://cdn.bootcss.com/mathjax/2.7.1/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:" && /^https?:/.test(src))
      src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
